### ISSUE TITLE
dbt does not cleanup on Athena error and subsequent retries fail

### ISSUE THREAD

**Issue by juliansteger-sc** [dbt-athena/dbt-athena#371]
We're using the `ha=true` variant with `s3_data_naming = schema_table_unique`.

Sometimes we get a `Query exhausted resources at this scale factor.` but data was already partially written to S3.

In subsequent tries we then receive something like `HIVE_PATH_ALREADY_EXISTS: Target directory for table 'my_table__ha' already exists:` - I think it is because dbt-athena does not cleanup the temporary location when there was an error.

**Comment by nicor88** [comment]
Let me get this right @juliansteger-sc table in the "glue" catalog does not exist - due to query failure on CTA - but path exists in s3? If so can be easy as add a delete call in the s3 path used to write the tmp ha table.

**Comment by juliansteger-sc** [comment]
correct, there is data in S3 (so Athena apparently managed to write to S3), but the query eventually failed with above error. It happens very rarely, so its hard to reproduce, but this is what I could see in Athena query history. (And it probably applies to any 500 error, when Athena already put some data to S3)

From what I can tell, retrying is managed via `tenacity.Retrying`, and at this point we don't have access to the S3 location anymore (its part of the query string). We could probably extract the s3 location from the query string but maybe there's a better way

**Comment by nicor88** [comment]
@juliansteger-sc I see what it's happening - you have retry >0, which means the same exact location is used by dbt-athena. Did you consider using retry=0 in dbt-athena and adding retries to your orchestrator (airflow/step functions or whatever you use)? doing so a new location path should be generated and the query should succeed at some point in the following attempts.

This specific issue is not super trivial to fix, as the location it's generated via a UUID, therefore we should implement a retry mechanism that has knowledge of the location - following more on this, I believe that we can catch and collect the result of the query that we send, and in case we receive HIVE_PATH_ALREADY_EXISTS, we could parse the s3 location. Of course it's more tricky.

**Comment by juliansteger-sc** [comment]
I'm not sure we should generally prune on `HIVE_PATH_ALREADY_EXISTS`, there might be cases when its valid to stop processing here. 
We only need to do proper cleanup after we encountered a [system error](https://docs.aws.amazon.com/athena/latest/ug/error-reference.html) and then retry.

So when we encounter an error with error category 1 (`SYSTEM`) (can check this via `aws athena get-query-execution` / boto3), we need to cleanup the temporary directory. 

So with the current retry mechanism since we need to proceed like this

If there's an error during execution
1. catch the error, and parse the query `execution_id` uuid from the error
2. call `GetQueryExecution` API Endpoint and fetch status / error code
3. if error code is 1, parse temporary location from query string and prune it, for other error codes do nothing

**Comment by nicor88** [comment]
@juliansteger-sc you right make sense in specific "cases" to do a proper clean up - feel free to give a shot at what you proposed.

Going back to the route case - or quick workaround - did you try retry=0 and adding orchestrator retries? should do the job.

**Comment by juliansteger-sc** [comment]
> Going back to the route case - or quick workaround - did you try retry=0 and adding orchestrator retries? should do the job.

Technically letting the orchestrator retry would work, but we would rather not redo the whole `dbt run` as there are some more expensive models involved. 

Also `retry > 0` works fine in cases when there is an Athena 5xx error and no data has been written yet (which is most of the cases).

**Comment by tomholt1** [comment]
Hi both. I'm wondering if you ever found a solution to this @juliansteger-sc ? We are in a similar situation here with getting these errors. I'd be keep to hear if you ever got a workaround for this, or if you managed to retry in any other way? Thanks

**Comment by juliansteger-sc** [comment]
hey @tomholt1, no we did not (yet), right now we just accept the occasional alerts
